### ISSUE TITLE
Operator "+" not supported for types "list[Tuple[Literal[''], Literal['foo']]]" and "list[Tuple[Literal['bar'], Literal['baz']]]"

### ISSUE THREAD

**Issue by federicobond** [microsoft/pyright#781]
The following code fails to typecheck:

```python
[('', 'foo')] + [('bar', 'baz')]
```
It gives the error:

> Operator "+" not supported for types "list[Tuple[Literal[''], Literal['foo']]]" and "list[Tuple[Literal['bar'], Literal['baz']]]"

I am running Pyright in VS Code. Might be a similar case to https://github.com/microsoft/pyright/issues/441

**Comment by federicobond** [comment]
It should also typecheck correctly between tuples with `str` and `Literal['...']` elements like this example:

```python
a = str("")
[("foo", "bar")] + [(a, a)]
```

**Comment by erictraut** [comment]
This is intended behavior. 

When Pyright sees the expression `('', 'foo')`, it needs to infer its type if there is no annotation provided. The following types would arguably be valid for this expression:
```
Tuple[Literal[''], Literal['foo']]
Tuple[str, str]
Tuple[str, ...]
Tuple[Any, ...]
```
I've listed those in order of specific to general. Without additional guidance from an annotation, Pyright uses heuristics to infer the type. In the case of tuples, it uses the most specific type possible, since that generally provides the best type safety.

Similarly, when Pyright encounters a list expression (e.g. `[1, "hi"]` or `[3.4, 2.3]`) without any type annotation, it needs to infer the type. In the case of lists, it uses the following heuristic:
1. If all of the elements in the list are the same type T, assume the type of the list is `List[T]`.
2. If the element types don't match and the `strictListInference` configuration setting is set to `true`, assume the type of the list is `List[Union[A, B, C, ...]]`.
3. If `strictListInference` is `false` (which it is by default), assume the type of the list is `List[Any]`.

Applying these rules to your example:
The type of expression `('', 'foo')` is `Tuple[Literal[''], Literal['foo']]`
The type of expression `[('', 'foo')] is `List[Tuple[Literal[''], Literal['foo']]]`
Similarly, the type of expression `[('bar', 'baz')]` is `List[Tuple[Literal['bar'], Literal['baz']]`
If you attempt to add these together, the `__add__` method on the `list` class indicates that there is a type violation because the types of the lists don't match.

The correct workaround in this case is to do the following:
```
StringListTuple = List[Tuple[str, str]]
a: StringListTuple = [('', 'foo')]
a + [('bar', 'baz')]
```
Note that I've created a type alias `StringListTuple` for convenience here.

**Comment by federicobond** [comment]
Oh ok, this case is very similar to #441 which is marked as fixed, hence the confusion. The original code from #441 does not typecheck with the latest version so I assume the decision has been reversed at some point.

The result is bit unexpected since only tuples work this way:

```python
["foo"]  # has type list[str]

["foo"] + ["bar"]  # ok

[("foo",)] + [("bar",)]  # type error because ("foo",) has type Tuple[Literal['foo']] 
```

It would be useful to document this behaviour somewhere.
